File: apps/android/app/src/main/java/com/rayact/engine/RayactEngineHostCallbacks.kt
package com.rayact.engine

interface RayactEngineHostCallbacks {
    fun requestNewSurface(): Int
    fun rootSurfaceId(): Int
    fun topSurfaceId(): Int
    fun releaseTopSurface()
    fun releaseSurface(surfaceId: Int)
    fun orderSurfaces(surfaceIds: IntArray)
    fun finishActivity()
    fun toggleDevMenu()
    fun requestRenderFrame()
    fun sendDevtoolsMessage(message: String)

    /**
     * Stop this session's render thread synchronously (blocks until it has
     * quit). Called by native code before tearing down the graphics device
     * when another session steals the graphics lease.
     */
    fun stopRenderScheduler()
    fun showSoftKeyboard(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Boolean,
        secure: Boolean,
        imeAction: String,
        autoCapitalize: String,
        contextMenuHidden: Boolean
    )
    fun switchIme(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Boolean,
        secure: Boolean,
        imeAction: String,
        autoCapitalize: String,
        contextMenuHidden: Boolean
    )
    fun hideSoftKeyboard()
    fun updateImeState(
        nodeId: Int,
        selectionStart: Int,
        selectionEnd: Int,
        composingStart: Int,
        composingEnd: Int,
        text: String?
    )
    fun copyToClipboard(text: String)
    fun readClipboard(): String
    fun performHapticFeedback()
}

File: apps/ios/RayactEngineHostCallbacks.swift
import Foundation

protocol RayactEngineHostCallbacks: AnyObject {
    func requestNewSurface() -> Int
    func rootSurfaceId() -> Int
    func topSurfaceId() -> Int
    func releaseTopSurface()
    func releaseSurface(surfaceId: Int)
    func orderSurfaces(surfaceIds: [Int])
    func finishActivity()
    func toggleDevMenu()
    func requestRenderFrame()
    func sendDevtoolsMessage(_ message: String)
    func stopRenderScheduler()
    func showSoftKeyboard(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Bool,
        secure: Bool,
        imeAction: String,
        autoCapitalize: String,
        contextMenuHidden: Bool
    )
    func switchIme(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Bool,
        secure: Bool,
        imeAction: String,
        autoCapitalize: String,
        contextMenuHidden: Bool
    )
    func hideSoftKeyboard()
    func updateImeState(
        nodeId: Int,
        selectionStart: Int,
        selectionEnd: Int,
        composingStart: Int,
        composingEnd: Int,
        text: String?
    )
    func copyToClipboard(text: String)
    func readClipboard() -> String
    func performHapticFeedback()
}
